rcl/src/rcl/security.c

```diff
@@ -0,0 +1,310 @@
+
+  if (!use_security) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  get_env_error_str = rcutils_get_env(ROS_SECURITY_ENABLE_VAR_NAME, &ros_security_enable);
+  if (NULL != get_env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var '" RCUTILS_STRINGIFY(ROS_SECURITY_ENABLE_VAR_NAME) "': %s\n",
+      get_env_error_str);
+    return RCL_RET_ERROR;
+  }
+
+  *use_security = (0 == strcmp(ros_security_enable, "true"));
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_get_enforcement_policy(rmw_security_enforcement_policy_t * policy)
+{
+  const char * ros_enforce_security = NULL;
+  const char * get_env_error_str = NULL;
+
+  if (!policy) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  get_env_error_str = rcutils_get_env(ROS_SECURITY_STRATEGY_VAR_NAME, &ros_enforce_security);
+  if (NULL != get_env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var '" RCUTILS_STRINGIFY(ROS_SECURITY_STRATEGY_VAR_NAME) "': %s\n",
+      get_env_error_str);
+    return RCL_RET_ERROR;
+  }
+
+  *policy = (0 == strcmp(ros_enforce_security, "Enforce")) ?
+    RMW_SECURITY_ENFORCEMENT_ENFORCE : RMW_SECURITY_ENFORCEMENT_PERMISSIVE;
+  return RCL_RET_OK;
+}
+
+/**
+ * A security lookup function takes in the node's name, namespace, a security root directory and an allocator;
+ *  It returns the relevant information required to load the security credentials,
+ *   which is currently a path to a directory on the filesystem containing DDS Security permission files.
+ */
+typedef char * (* security_lookup_fn_t) (
+  const char * node_name,
+  const char * node_namespace,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+char * exact_match_lookup(
+  const char * node_name,
+  const char * node_namespace,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+char * prefix_match_lookup(
+  const char * node_name,
+  const char * node_namespace,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+security_lookup_fn_t g_security_lookup_fns[] = {
+  NULL,
+  exact_match_lookup,
+  prefix_match_lookup,
+};
+
+typedef enum ros_security_lookup_type_e
+{
+  ROS_SECURITY_LOOKUP_NODE_OVERRIDE = 0,
+  ROS_SECURITY_LOOKUP_MATCH_EXACT = 1,
+  ROS_SECURITY_LOOKUP_MATCH_PREFIX = 2,
+} ros_security_lookup_type_t;
+
+char * g_security_lookup_type_strings[] = {
+  "NODE_OVERRIDE",
+  "MATCH_EXACT",
+  "MATCH_PREFIX"
+};
+
+/// Return the directory whose name most closely matches node_name (longest-prefix match),
+/// scanning under base_dir.
+/**
+ * By using a prefix match, a node named e.g. "my_node_123" will be able to load and use the
+ * directory "my_node" if no better match exists.
+ * \param[in] base_dir
+ * \param[in] node_name
+ * \param[out] matched_name must be a valid memory address allocated with at least
+ * _TINYDIR_FILENAME_MAX characters.
+ * \return true if a match was found
+ */
+static bool get_best_matching_directory(
```

Comment:
Is this file a copy of `security_directory.c` ?
Is there a reason to duplicate the files?

If no need to keep the old file, could we just rename it and preserve history instead

Reply:
Oh, I thought I renamed (I actually renamed in the `CMakeLists.txt`, and `security_directory` is not compiled more).
Don't remember what I did, I will try to clean the history.

Reply:
I've cleaned the history now.